### PR TITLE
introduce RouteSetProgressDelegate

### PR DESCRIPTION
**Pull request by dzinad**

### CODE REVIEW

**Comment by dzinad** [comment]
Just a draft to agree on the API and to see if that solution would work. I'll add tests, docs, etc. later.

**Comment by Guardiola31337** [comment]
> Just a draft to agree on the API and to see if that solution would work.

Took this for a quick spin downstream and we should find a way to provide the `NavigationSessionState` through these callbacks and somehow fire them off only around that time, currently `onStarted` / `onSuccess` are fired off on every `setRoutes` which is confusing as a client, as you're only interested in the very first one (the one that ends up changing the navigation session state) and any other (re-routes, refresh requests, etc.) are irrelevant.

```
// MapboxNavigation starts a session in Idle
I/System.out: DEBUG NavigationSessionStateObserver#onNavigationSessionStateChanged(com.mapbox.navigation.core.trip.session.NavigationSessionState$Idle@e29fdec)

// MapboxNavigation transitions to Free Drive (empty setRoute)
I/System.out: DEBUG RouteSetProgressDelegate#onStarted
I/System.out: DEBUG NavigationSessionStateObserver#onNavigationSessionStateChanged(FreeDrive(sessionId=ac754f8f-9c76-4c4d-84ef-9c7e79a84e61))
I/System.out: DEBUG RouteSetProgressDelegate#onSuccess

// MapboxNavigation transitions to Active Guidance (non-empty setRoute)
I/System.out: DEBUG RouteSetProgressDelegate#onStarted
I/System.out: DEBUG NavigationSessionStateObserver#onNavigationSessionStateChanged(ActiveGuidance(sessionId=0fea21dc-cfc6-4a38-897e-15dd06ed6f2f))
I/System.out: DEBUG RouteSetProgressDelegate#onSuccess

//
// All good up to here
//

// Off-route - this is irrelevant for clients
I/System.out: DEBUG RouteSetProgressDelegate#onStarted
I/System.out: DEBUG RouteSetProgressDelegate#onSuccess

// MapboxNavigation transitions to Free Drive (empty setRoute)
I/System.out: DEBUG RouteSetProgressDelegate#onStarted
I/System.out: DEBUG NavigationSessionStateObserver#onNavigationSessionStateChanged(FreeDrive(sessionId=6499947c-1956-4d7a-87c6-40ca78b896f3))
I/System.out: DEBUG RouteSetProgressDelegate#onSuccess

// MapboxNavigation stops the session (Idle)
I/System.out: DEBUG NavigationSessionStateObserver#onNavigationSessionStateChanged(com.mapbox.navigation.core.trip.session.NavigationSessionState$Idle@e29fdec)
```

If we go this path (as it's currently suggested and above-described), clients would need to introduce some sort of synchronization between `RouteSetProgressDelegate` and `NavigationSessionStateObserver` (potentially even the `RoutesObserver`) to react accordingly as we used to with only the `NavigationSessionStateObserver` https://github.com/mapbox/mapbox-navigation-android/pull/5938 which would definitely complicate the integration and may introduce other issues (races).

@dzinad any ideas to cover ^^?

**Comment by dzinad** [comment]
I see. What about sth like this?
We introduce a "preview" of `NavigationSession#setRoutes`. It does not change the state but it tells if the state would be changed (considering the provided routes are accepted by the navigator) and, if so, from which one to which one.  That information would then be passed to `onStarted`. After that, we would pass the actual final state to `onSuccess`/`onFailure`. Would that work? Would it be easy to use? I imagine sth like this:
```
        val delegate = object : RouteSetProgressDelegate {
            override fun onStarted(prevState: NavigationSessionState, potentialNewState: NavigationSessionState) {
                if (prevState != potentialNewState && prevState is Idle) {
                    // start recording
                }
            }

            override fun onSuccess(actualNewState: NavigationSessionState) {
                if (actualNewState is Idle) {
                    // stop recording
                }
            }

            override fun onFailure(actualNewState: NavigationSessionState) {
                if (actualNewState is Idle) {
                    // stop recording
                }
            }
        }
```

Actually, it looks like we don't need to distinguish success from failure, we just need the actual session state.
Or we can event simplify the usage: we can check the ifs from our side and only fire the callback methods: `onShouldStartRecording` and `onShouldStopRecording`. Depends on how much flexibility we want.
WDYT?

**Comment by dzinad** [comment]
I updated the API. @Guardiola31337 could you please have a look and give it a test in Copilot?

**Comment by Guardiola31337** [comment]
Took this for another spin and it seems that current implementation wouldn't completely work 👀 

```
// MapboxNavigation starts a session in Idle
2022-07-13 13:02:04.535 20403-20403/com.mapbox.navigation.copilot.v2.android I/System.out: PGS onNavigationSessionStateChanged(com.mapbox.navigation.core.trip.session.NavigationSessionState$Idle@cd1a5c6)

// MapboxNavigation transitions to Free Drive (empty setRoute) we're missing onShouldStopRecording and onShouldStartRecording
2022-07-13 13:02:09.516 20403-20403/com.mapbox.navigation.copilot.v2.android I/System.out: PGS onNavigationSessionStateChanged(FreeDrive(sessionId=20457501-efdb-45f9-84d5-37a67ff3f34a))

// MapboxNavigation transitions to Active Guidance (non-empty setRoute)
2022-07-13 13:02:25.784 20403-20403/com.mapbox.navigation.copilot.v2.android I/System.out: PGS onShouldStopRecording(FreeDrive(sessionId=20457501-efdb-45f9-84d5-37a67ff3f34a))
2022-07-13 13:02:25.784 20403-20403/com.mapbox.navigation.copilot.v2.android I/System.out: PGS onShouldStartRecording(ActiveGuidance(sessionId=7c8d6452-cac6-4bdd-8be8-eb4d7fe4e328))
2022-07-13 13:02:25.793 20403-20403/com.mapbox.navigation.copilot.v2.android I/System.out: PGS onNavigationSessionStateChanged(ActiveGuidance(sessionId=0e22d69e-416d-4f7b-9731-372a735d6f85))

// MapboxNavigation transitions to Free Drive (empty setRoute)
2022-07-13 13:04:54.111 20403-20403/com.mapbox.navigation.copilot.v2.android I/System.out: PGS onShouldStopRecording(ActiveGuidance(sessionId=0e22d69e-416d-4f7b-9731-372a735d6f85))
2022-07-13 13:04:54.111 20403-20403/com.mapbox.navigation.copilot.v2.android I/System.out: PGS onShouldStartRecording(FreeDrive(sessionId=cb4f7006-03f5-4772-98aa-4ddd85375cd8))
2022-07-13 13:04:54.134 20403-20403/com.mapbox.navigation.copilot.v2.android I/System.out: PGS onNavigationSessionStateChanged(FreeDrive(sessionId=325783b8-0191-4152-8806-447a9647e6d0))

// MapboxNavigation stops the session (Idle) we're missing onShouldStopRecording and onShouldStartRecording
2022-07-13 13:05:01.128 20403-20403/com.mapbox.navigation.copilot.v2.android I/System.out: PGS onNavigationSessionStateChanged(com.mapbox.navigation.core.trip.session.NavigationSessionState$Idle@cd1a5c6)
```

It seems that we're still missing the `Idle` transitions. I believe that if we cover above missing callbacks we should be good.

**Comment by dzinad** [comment]
I also reverted https://github.com/mapbox/mapbox-navigation-android/pull/5938. Not sure if this is the correct place, but it makes sense to do these things together because otherwise we have a bug (either https://github.com/mapbox/mapbox-navigation-android/issues/6023 or absence of first SET_ROUTE event).

**Comment by Guardiola31337** [comment]
Tested latest changes and now we're getting 👇 

```
// MapboxNavigation starts a session in Idle
2022-07-14 11:56:53.116 7622-7622/com.mapbox.navigation.copilot.v2.android I/System.out: DEBUG onNavigationSessionStateChanged(com.mapbox.navigation.core.trip.session.NavigationSessionState$Idle@b59e9a8)

// MapboxNavigation transitions to Free Drive (empty setRoute)
2022-07-14 11:56:58.590 7622-7622/com.mapbox.navigation.copilot.v2.android I/System.out: DEBUG onShouldStartRecording(FreeDrive(sessionId=ba8d1fcf-3ba3-4279-b53c-7907c27620c9))
2022-07-14 11:56:58.626 7622-7622/com.mapbox.navigation.copilot.v2.android I/System.out: DEBUG onNavigationSessionStateChanged(FreeDrive(sessionId=ba8d1fcf-3ba3-4279-b53c-7907c27620c9))

// MapboxNavigation transitions to Active Guidance (non-empty setRoute)
2022-07-14 11:57:14.186 7622-7622/com.mapbox.navigation.copilot.v2.android I/System.out: DEBUG onShouldStopRecording(FreeDrive(sessionId=ba8d1fcf-3ba3-4279-b53c-7907c27620c9))
2022-07-14 11:57:14.186 7622-7622/com.mapbox.navigation.copilot.v2.android I/System.out: DEBUG onShouldStartRecording(ActiveGuidance(sessionId=244ee9f5-5aeb-487d-adc7-0083d9263445))
2022-07-14 11:57:14.210 7622-7622/com.mapbox.navigation.copilot.v2.android I/System.out: DEBUG onNavigationSessionStateChanged(ActiveGuidance(sessionId=17538894-b64c-4277-8c14-9375b8ead3f2))

// MapboxNavigation transitions to Free Drive (empty setRoute)
2022-07-14 11:57:41.154 7622-7622/com.mapbox.navigation.copilot.v2.android I/System.out: DEBUG onShouldStopRecording(ActiveGuidance(sessionId=244ee9f5-5aeb-487d-adc7-0083d9263445))
2022-07-14 11:57:41.154 7622-7622/com.mapbox.navigation.copilot.v2.android I/System.out: DEBUG onShouldStartRecording(FreeDrive(sessionId=16330ee1-7e47-4223-9429-2682aae2f191))
2022-07-14 11:57:41.179 7622-7622/com.mapbox.navigation.copilot.v2.android I/System.out: DEBUG onNavigationSessionStateChanged(FreeDrive(sessionId=4b35d992-3140-47c3-b628-0b0338a37c0c))

// MapboxNavigation stops the session (Idle)
2022-07-14 11:58:15.756 7622-7622/com.mapbox.navigation.copilot.v2.android I/System.out: DEBUG onShouldStopRecording(FreeDrive(sessionId=16330ee1-7e47-4223-9429-2682aae2f191))
2022-07-14 11:58:15.785 7622-7622/com.mapbox.navigation.copilot.v2.android I/System.out: DEBUG onNavigationSessionStateChanged(com.mapbox.navigation.core.trip.session.NavigationSessionState$Idle@b59e9a8)
```

Which seems fine but want to make sure that not reporting `Idle` was by design. Is it guaranteed that now the callbacks don't return `Idle` in any circumstance?

Also wondering how clients would implement the error path from their side. What happens in there's an error?

```
// MapboxNavigation starts a session in Idle
2022-07-14 11:56:53.116 7622-7622/com.mapbox.navigation.copilot.v2.android I/System.out: PGS onNavigationSessionStateChanged(com.mapbox.navigation.core.trip.session.NavigationSessionState$Idle@b59e9a8)

// MapboxNavigation transitions to Free Drive (empty setRoute)
2022-07-14 11:56:58.590 7622-7622/com.mapbox.navigation.copilot.v2.android I/System.out: PGS onShouldStartRecording(FreeDrive(sessionId=ba8d1fcf-3ba3-4279-b53c-7907c27620c9))
2022-07-14 11:56:58.626 7622-7622/com.mapbox.navigation.copilot.v2.android I/System.out: PGS onNavigationSessionStateChanged(FreeDrive(sessionId=ba8d1fcf-3ba3-4279-b53c-7907c27620c9))

// MapboxNavigation transitions to Active Guidance (non-empty setRoute)
2022-07-14 11:57:14.186 7622-7622/com.mapbox.navigation.copilot.v2.android I/System.out: PGS onShouldStopRecording(FreeDrive(sessionId=ba8d1fcf-3ba3-4279-b53c-7907c27620c9))
2022-07-14 11:57:14.186 7622-7622/com.mapbox.navigation.copilot.v2.android I/System.out: PGS onShouldStartRecording(ActiveGuidance(sessionId=244ee9f5-5aeb-487d-adc7-0083d9263445))

// What happens in case of error?

// onShouldStopRecording followed by onNavigationSessionStateChanged?
onShouldStopRecording(ActiveGuidance(sessionId=244ee9f5-5aeb-487d-adc7-0083d9263445))
onNavigationSessionStateChanged(FreeDrive(sessionId=<new-session-id>))

// onShouldStopRecording only?
onShouldStopRecording(ActiveGuidance(sessionId=244ee9f5-5aeb-487d-adc7-0083d9263445))
```

Could you clarify when you have a chance? Have you tested this scenario?

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/HistoryRecordingStateChangeObserver.kt [hunk 1]
```diff
@@ -0,0 +1,15 @@
+package com.mapbox.navigation.core
+
+import androidx.annotation.UiThread
+import com.mapbox.navigation.base.route.NavigationRoute
+import com.mapbox.navigation.core.trip.session.NavigationSessionState
+
+@UiThread
+interface HistoryRecordingStateChangeObserver {
```
Comment (LukasPaczos) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/HistoryRecordingStateChangeObserver.kt:8]: This should be in an internal package or tagged as experimental.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt [hunk 2]
```diff
@@ -840,6 +846,15 @@ class MapboxNavigation @VisibleForTesting internal constructor(
...
     }
 
+    // call before starting session - to docs
+    fun registerHistoryRecordingStateChangeObserver(observer: HistoryRecordingStateChangeObserver) {
+        historyRecordingStateHandler.registerHistoryRecordingStateChangeObserver(observer)
+    }
+
+    fun unregisterHistoryRecordingStateChangeObserver(observer: HistoryRecordingStateChangeObserver) {
+        historyRecordingStateHandler.unregisterHistoryRecordingStateChangeObserver(observer)
+    }
```
Comment (LukasPaczos) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt:856]: These should rather be extensions on `MapboxNavigation` exposed from `internal` package, or at least tagged with experimental.

**Review by LukasPaczos — COMMENTED** [review]
I don't have any concerns on the billing side, I don't see any sessions being started twice. I'm not sure whether we need to deliver the session type instance and effectively generate an invalid session ID for copilot, but we can address that at any time.

If we are planning to make something similar public, possibly we could lean towards a `HistoryRecorderOptions#recordEachSession` and manage the naming of files and IDs internally. Could copilot use this type of implicit recording management as well?

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt (hunk 2 above)
Comment (dzinad) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt:856]: https://github.com/mapbox/mapbox-navigation-android/pull/6074

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/HistoryRecordingStateChangeObserver.kt (hunk 1 above)
Comment (dzinad) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/HistoryRecordingStateChangeObserver.kt:8]: https://github.com/mapbox/mapbox-navigation-android/pull/6074

**Comment by dzinad** [comment]
> I don't have any concerns on the billing side, I don't see any sessions being started twice. I'm not sure whether we need to deliver the session type instance and effectively generate an invalid session ID for copilot, but we can address that at any time.
If we are planning to make something similar public, possibly we could lean towards a HistoryRecorderOptions#recordEachSession and manage the naming of files and IDs internally. Could copilot use this type of implicit recording management as well?

I've redone it a bit: https://github.com/mapbox/mapbox-navigation-android/pull/6074

**Pull request closed without merging**